cmd/esc/cli/env.go

```diff
@@ -65,49 +66,186 @@ func newEnvCmd(esc *escCommand) *cobra.Command {
 	}
 
 	envName, version, hasSep := strings.Cut(envNameAndVersion, "@")
 	if !hasSep {
 		envName, version, _ = strings.Cut(envNameAndVersion, ":")
 	}
 
-	if rel != nil && envName == "" && version != "" {
-		orgName, envName = rel.orgName, rel.envName
+	return environmentRef{
+		orgName:          orgName,
+		projectName:      projectName,
+		envName:          envName,
+		version:          version,
+		hasAmbiguousPath: hasAmbiguousPath,
 	}
+}
+
+// If a non-nil environmentRef is provided, default to its values if an environment name and version are not specified
+func (cmd *envCommand) getEnvRef(refString string, rel *environmentRef) (ref environmentRef) {
+	envRef := cmd.parseRef(refString)
 
-	return environmentRef{orgName, envName, version}
+	if rel != nil && envRef.envName == "" && envRef.version != "" {
+		envRef.orgName = rel.orgName
+		envRef.projectName = rel.projectName
+		envRef.envName = rel.envName
+	}
+
+	return envRef
 }
 
-func (cmd *envCommand) getEnvRef(args []string) (ref environmentRef, rest []string, err error) {
+// Get an environment reference when creating a new environment
+// If the given path is ambiguous, we need to make an additional API call to disambiguate
+func (cmd *envCommand) getNewEnvRef(
+	ctx context.Context,
+	args []string,
+) (ref environmentRef, rest []string, err error) {
 	if cmd.envNameFlag == "" {
 		if len(args) == 0 {
 			return environmentRef{}, nil, fmt.Errorf("no environment name specified")
 		}
 		cmd.envNameFlag, args = args[0], args[1:]
 	}
 
-	return cmd.getRelativeEnvRef(cmd.envNameFlag, nil), args, nil
+	ref = cmd.getEnvRef(cmd.envNameFlag, nil)
+
+	if !ref.hasAmbiguousPath {
+		return ref, args, nil
+	}
+
+	// Check if user is part of the organization from legacy path pattern <org-name>/default/<environment-name>
+	legacyRef := environmentRef{
+		orgName:          ref.projectName,
+		projectName:      client.DefaultProject,
+		envName:          ref.envName,
+		version:          ref.version,
+		hasAmbiguousPath: ref.hasAmbiguousPath,
+	}
+
+	useLegacyPath := false
+	_, orgs, _, _ := cmd.esc.client.GetPulumiAccountDetails(ctx)
+	for _, org := range orgs {
+		if strings.EqualFold(legacyRef.orgName, org) {
+			useLegacyPath = true
+			break
+		}
+	}
+
+	if useLegacyPath {
+		return legacyRef, args, nil
+	}
+
+	return ref, args, nil
+}
+
+// Get an environment reference for an existing environment
+// If the given path is ambiguous, we need to make additional API calls to disambiguate
+func (cmd *envCommand) getExistingEnvRef(
+	ctx context.Context,
+	args []string,
+) (ref environmentRef, rest []string, err error) {
+	if cmd.envNameFlag == "" {
+		if len(args) == 0 {
+			return environmentRef{}, nil, fmt.Errorf("no environment name specified")
+		}
+		cmd.envNameFlag, args = args[0], args[1:]
+	}
+
+	envRef, err := cmd.getExistingEnvRefWithRelative(ctx, cmd.envNameFlag, nil)
+
+	return envRef, args, err
+}
+
+func (cmd *envCommand) getExistingEnvRefWithRelative(
```

Comment:
The naming here is a bit confusing imo I think the original intention of the ref passed in was to be a way to resolve a env reference string without an env name but a version specified. WithRelative is a bit unclear how the relative is being used. Even the existing function name is a bit unclear to me since the reference returned is not always relative to the reference passed in (e.g. if the ref string contains an environment name)

Reply:
Yah I can see it's confusing. `getExistingEnvRef` refers to a reference for an existing env, as opposed to `getNewEnvRef` which refers to a reference for a new env that is about to be created. (We differentiate between these 2 cases because of differing logic for checking for the legacy pattern).

Would `getRelativeExistingEnvRef` be better? or `getExistingEnvRefWithFallback`?

Reply:
definitely not a blocker so feel free to continue without changing but I think `getExistingEnvRefWithFallback` might be clearer